src/coreclr/jit/morph.cpp

```diff
@@ -13614,53 +13642,57 @@ void Compiler::fgMorphBlocks()
         lvSetMinOptsDoNotEnreg();
     }
 
-    /*-------------------------------------------------------------------------
-     * Process all basic blocks in the function
-     */
-
-    BasicBlock* block = fgFirstBB;
-    noway_assert(block);
-
-    do
+    // Morph all blocks. If we aren't then we just morph in normal bbNext order.
+    //
+    if (!optLocalAssertionProp)
     {
-#ifdef DEBUG
-        if (verbose)
+        // Note morph can add blocks downstream from the current block,
+        // and alter (but not null out) the current block's bbNext;
+        // this iterator ensures they all get visited.
+        //
+        for (BasicBlock* block : Blocks())
         {
-            printf("\nMorphing " FMT_BB " of '%s'\n", block->bbNum, info.compFullName);
+            fgMorphBlock(block);
         }
-#endif
+    }
+    else
+    {
+        // We are optimizing. Process in RPO.
+        //
+        fgRenumberBlocks();
+        EnsureBasicBlockEpoch();
+        fgComputeEnterBlocksSet();
+        fgDfsReversePostorder();
```

Comment:
I'm curious why we need to do so much work to set up for the RPO traversal. SSA's `TopologicalSort` does not do any of this work; it simply starts at `fgFirstBB` (well, the BB root, but then the original `fgFirstBB` very soon after that). Would it be incorrect to do the same here?

Another idea: can we avoid the two step iteration process, and instead have a version of `fgDfsReversePostorder` that just invokes a callback instead of recording the order into ambient state? Would it make any difference? I'm not totally sure where the throughput cost comes from in this change.

Reply:
> I'm curious why we need to do so much work to set up for the RPO traversa

We can try and do all this on the fly, that's more or less what I was suggesting above:

>  > we could do a dynamic evolution of the RPO by turning this into a worklist driven traversal (similar to how the importer runs).

> I'm not totally sure where the throughput cost comes from in this change.

I don't understand it either, seems like this latest version should (aside from building the RPO) be fairly efficient. Will have to take a look.

Reply:
> Another idea: can we avoid the two step iteration process, and instead have a version of `fgDfsReversePostorder` that just invokes a callback instead of recording the order into ambient state? Would it make any difference? I'm not totally sure where the throughput cost comes from in this change.

I suspect this is tricky for phases like morph that also can alter the flow graph.